### PR TITLE
Add Seq Packing in NeMo / Neva2

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/multimodal/data/energon/config.py [hunk 1]
```diff
@@ -43,6 +46,15 @@
     loss_mask: torch.Tensor = field(default_factory=lambda: torch.empty(0, dtype=torch.float))
 
 
+@dataclass
+class PackedImageTextSample(ImageTextSample):
+    '''Sample type for packed image text sample'''
+
+    __restore_key__: Tuple[Union[str, int, tuple], ...] = ()
+    position_ids: torch.Tensor = field(default_factory=lambda: torch.empty(0, dtype=torch.float))
+    packed_seq_params: PackedSeqParams = field(default_factory=lambda: PackedSeqParams())
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/multimodal/data/energon/config.py:55]: can you resolve this?

File: nemo/collections/multimodal/data/energon/task_encoder.py [hunk 2]
```diff
@@ -54,7 +61,15 @@ class MultiModalTaskEncoder(
...
     """
 
-    def __init__(self, tokenizer, image_processor, multimodal_sample_config):
+    def __init__(
+        self,
+        tokenizer,
+        image_processor,
+        multimodal_sample_config,
+        packed_sequence=False,
+        packing_seq_length=4096,
```
Comment (cuichenx) [inline nemo/collections/multimodal/data/energon/task_encoder.py:70]: can you use consistent naming as fine_tuning.py? i.e. 
use one variable `packed_sequence_size`, and `packed_sequence_size=-1` indicates not using packed sequence

File: nemo/collections/vlm/neva/data/config.py [hunk 3]
```diff
@@ -31,15 +31,15 @@ class DataConfig:
 @dataclass
 class ImageDataConfig(DataConfig):
     media_type: str = "image"
-    media_token: MultiModalToken = ImageToken
+    media_token: MultiModalToken = ImageToken()
```
Comment (cuichenx) [inline nemo/collections/vlm/neva/data/config.py:34]: can you double check if dataclass can initialize a class instance as default value? I thought you have to use default factory

File: nemo/lightning/megatron_parallel.py [hunk 4]
```diff
@@ -1711,7 +1711,10 @@ def masked_token_loss(tensor: Tensor, mask: Tensor):
     """
     losses = tensor.float()
     loss_mask = mask.view(-1).float()
-    loss = torch.sum(losses.view(-1) * loss_mask) / loss_mask.sum()  # sequence level nll
+    num_valid_tokens = loss_mask.sum()
+    if num_valid_tokens < 0.5:  # no valid tokens
```
Comment (cuichenx) [inline nemo/lightning/megatron_parallel.py:1715]: can you explain a bit more when this is the case? is this only valid for neva?

Comment (yaoyu-33) [reply nemo/lightning/megatron_parallel.py:1715]: not only for neva, also for SFT. If the system and user prompt is very long, and predict answer only. After truncation from right, there might not be any answer/valid tokens.

File: nemo/collections/multimodal/data/energon/task_encoder.py (hunk 2 above)
Comment (yaoyu-33) [reply nemo/collections/multimodal/data/energon/task_encoder.py:70]: i kept `packed_sequence` here for consistency with other vlm data modules where we don't use packed_sequence_size

File: nemo/lightning/megatron_parallel.py (hunk 4 above)
Comment (cuichenx) [reply nemo/lightning/megatron_parallel.py:1715]: Usually we truncate the input/context and keep answer intact, so that wouldn't happen

Comment (yaoyu-33) [reply nemo/lightning/megatron_parallel.py:1715]: yep the logic is a bit different for vlm. We don't want to truncate from left.

**Review by cuichenx — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]
